[Bug]: Entire reward balance is not distributed to parties

### Problem encountered
in ST nightly run


I had a chat with Will about this issue,

In the case where a node operator delegates to another validator node, is that sometimes (because it depends on the order we iterate over this list in core) the node that delegated to another node will only get their rewards for what they self-delegated. We calculate the reward they should get for delegating to a different node but then never transfer it. So it stays in the reward accounts and thats why the test fails.


### Observed behaviour

reward % payouts to each party does not total 100







### Expected behaviour

As per the spec https://github.com/vegaprotocol/specs/blob/master/protocol/0056-REWA-rewards_overview.md

`Then the entire balance of each reward account is distributed to the parties with a non-zero reward metric for that reward type and market, pro-rata by their reward metric.`


I would expect 100% distribution of the reward balance

### Steps to reproduce

```bash
1. I've not been able to reliably reproduce this issue and only seen it in the overnight test failure
2. Either run the nightly run or run the following ST `test_rewards_after_rotation` on vega `develop` branch and hope that it produces the issue
3. Set the jenkins parameter `TEST_EXTRA_PYTEST_ARGS` with value `--count=15` to rerun the test 15 times  
4. Observe the error
`AssertionError: cannot complete test reward pool is not empty, need static account balances over rotation
assert 152122364848412 == 0`

And observing the `BUS_EVENT_TYPE_REWARD_PAYOUT_EVENT` shows the distribution percentage which when totalled with all payouts in that round does not equal 100%
```


### Software version

dev branch

### Failing test

test_rewards_after_rotation

### Jenkins run

https://jenkins.ops.vega.xyz/job/common/job/system-tests-wrapper/87983/testReport/junit/tests.wallet/vega_key_rotation_test/test_rewards_after_rotation/

### Configuration used

_No response_

### Relevant log output

```shell
{"id":"1093-35","block":"D1F90957ACF88020D146C0E7EC9740250588EA5EBB045A38EDD8AB8FF46EE619","type":"BUS_EVENT_TYPE_REWARD_PAYOUT_EVENT","rewardPayout":{"party":"6a9d3fc107441766f6958ebe11f35b2a01a793ca4683e5e6ac1bb482ee7b1323","epochSeq":"124","asset":"b4f2726571fbe8e33b442dc92ed2d7f0d810e21835b7371a7915a365f07ccd9b","amount":"161290322580645200","percentOfTotalReward":"16.12903225806452","timestamp":"1690683293175860828","rewardType":"ACCOUNT_TYPE_GLOBAL_REWARD","market":"!"},"version":1,"chainId":"testnet-002","txHash":"D1F90957ACF88020D146C0E7EC9740250588EA5EBB045A38EDD8AB8FF46EE619"}
{"id":"1093-36","block":"D1F90957ACF88020D146C0E7EC9740250588EA5EBB045A38EDD8AB8FF46EE619","type":"BUS_EVENT_TYPE_REWARD_PAYOUT_EVENT","rewardPayout":{"party":"7596f44300ac976cc10e9fed7c9a08ac4a43331ba462e1a1dc4ce3e9ca7048b6","epochSeq":"124","asset":"b4f2726571fbe8e33b442dc92ed2d7f0d810e21835b7371a7915a365f07ccd9b","amount":"69999999999999998","percentOfTotalReward":"7","timestamp":"1690683293175860828","rewardType":"ACCOUNT_TYPE_GLOBAL_REWARD","market":"!"},"version":1,"chainId":"testnet-002","txHash":"D1F90957ACF88020D146C0E7EC9740250588EA5EBB045A38EDD8AB8FF46EE619"}
{"id":"1093-37","block":"D1F90957ACF88020D146C0E7EC9740250588EA5EBB045A38EDD8AB8FF46EE619","type":"BUS_EVENT_TYPE_REWARD_PAYOUT_EVENT","rewardPayout":{"party":"93316f00cae3446c8c6aed9fc9a46348d0e5e325519358247ac489a28368e1f5","epochSeq":"124","asset":"b4f2726571fbe8e33b442dc92ed2d7f0d810e21835b7371a7915a365f07ccd9b","amount":"161290322580645200","percentOfTotalReward":"16.12903225806452","timestamp":"1690683293175860828","rewardType":"ACCOUNT_TYPE_GLOBAL_REWARD","market":"!"},"version":1,"chainId":"testnet-002","txHash":"D1F90957ACF88020D146C0E7EC9740250588EA5EBB045A38EDD8AB8FF46EE619"}
{"id":"1093-38","block":"D1F90957ACF88020D146C0E7EC9740250588EA5EBB045A38EDD8AB8FF46EE619","type":"BUS_EVENT_TYPE_REWARD_PAYOUT_EVENT","rewardPayout":{"party":"c15eb84d6fc0fc2c06fe9fcf0a5851bd2f12fa81369afcf1351f99ea672af8f2","epochSeq":"124","asset":"b4f2726571fbe8e33b442dc92ed2d7f0d810e21835b7371a7915a365f07ccd9b","amount":"161290322580645100","percentOfTotalReward":"16.12903225806451","timestamp":"1690683293175860828","rewardType":"ACCOUNT_TYPE_GLOBAL_REWARD","market":"!"},"version":1,"chainId":"testnet-002","txHash":"D1F90957ACF88020D146C0E7EC9740250588EA5EBB045A38EDD8AB8FF46EE619"}
{"id":"1093-39","block":"D1F90957ACF88020D146C0E7EC9740250588EA5EBB045A38EDD8AB8FF46EE619","type":"BUS_EVENT_TYPE_REWARD_PAYOUT_EVENT","rewardPayout":{"party":"d42c2465742056316c3469ca539a71e883e273a6e6f54d930d1eafa45cdf7bd4","epochSeq":"124","asset":"b4f2726571fbe8e33b442dc92ed2d7f0d810e21835b7371a7915a365f07ccd9b","amount":"161290322580645200","percentOfTotalReward":"16.12903225806452","timestamp":"1690683293175860828","rewardType":"ACCOUNT_TYPE_GLOBAL_REWARD","market":"!"},"version":1,"chainId":"testnet-002","txHash":"D1F90957ACF88020D146C0E7EC9740250588EA5EBB045A38EDD8AB8FF46EE619"}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -75,6 +75,7 @@
 - [8855](https://github.com/vegaprotocol/vega/issues/8855) - Preserve reference to parent market when restoring checkpoint data
 - [8909](https://github.com/vegaprotocol/vega/issues/8909) - initialise id generator for all branches of market state update
 - [8908](https://github.com/vegaprotocol/vega/issues/8908) - A rejected parent market should result in all successors getting rejected, too.
+- [8953](https://github.com/vegaprotocol/vega/issues/8953) - Fix reward distribution when validator delegate to other validators
 
 ## 0.72.1
 

--- a/core/rewards/staking_reward_calculator.go
+++ b/core/rewards/staking_reward_calculator.go
@@ -72,13 +72,13 @@ func calculateRewardsByStake(epochSeq, asset, accountID string, rewardBalance *n
 			logging.String("amountToKeepByValidator", amountToKeepByValidator.String()))
 
 		// check how much reward the validator can accept with the cap per participant
-		rewardForNode, ok := rewards[vd.NodeID]
+		rewardForNode, ok := rewards[vd.PubKey]
 		if !ok {
 			rewardForNode = num.UintZero()
 		}
 		// if there is no cap just add the total payout for the validator
 		if maxPayout.IsZero() {
-			rewards[vd.PubKey] = num.UintZero().Add(rewardForNode, amountToKeepByValidator)
+			rewards[vd.PubKey] = num.Sum(rewardForNode, amountToKeepByValidator)
 			totalRewardPayout.AddSum(amountToKeepByValidator)
 		} else {
 			balanceWithPayout := num.UintZero().Add(rewardForNode, amountToKeepByValidator)

--- a/core/rewards/staking_reward_calculator_test.go
+++ b/core/rewards/staking_reward_calculator_test.go
@@ -428,3 +428,86 @@ func testEarlyStopCalcRewardsMaxPayoutBreachedPartyCanTakeMore(t *testing.T) {
 	require.Equal(t, num.NewUint(207142), res.partyToAmount["node3"])
 	require.Equal(t, num.NewUint(999999), res.totalReward)
 }
+
+func TestValidatorIsNotBeingStarved(t *testing.T) {
+	delegatorShare, _ := num.DecimalFromString("0.3")
+	delegatorForVal1 := map[string]*num.Uint{}
+	delegatorForVal1["party1"] = num.NewUint(6000)
+	delegatorForVal1["party2"] = num.NewUint(3000)
+	delegatorForVal1["n2"] = num.NewUint(1000)
+	validator1 := &types.ValidatorData{
+		NodeID:            "node1",
+		PubKey:            "n1",
+		SelfStake:         num.UintZero(),
+		StakeByDelegators: num.NewUint(10000),
+		Delegators:        delegatorForVal1,
+	}
+	validator2 := &types.ValidatorData{
+		NodeID:            "node2",
+		PubKey:            "n2",
+		SelfStake:         num.NewUint(20000),
+		StakeByDelegators: num.UintZero(),
+		Delegators:        map[string]*num.Uint{},
+	}
+
+	delegatorForVal3 := map[string]*num.Uint{}
+	delegatorForVal3["party1"] = num.NewUint(40000)
+	validator3 := &types.ValidatorData{
+		NodeID:            "node3",
+		PubKey:            "n3",
+		SelfStake:         num.NewUint(30000),
+		StakeByDelegators: num.NewUint(40000),
+		Delegators:        delegatorForVal3,
+	}
+
+	validator4 := &types.ValidatorData{
+		NodeID:            "node4",
+		PubKey:            "n4",
+		SelfStake:         num.UintZero(),
+		StakeByDelegators: num.UintZero(),
+		Delegators:        map[string]*num.Uint{},
+	}
+	ctrl := gomock.NewController(t)
+	broker := bmocks.NewMockBroker(ctrl)
+	broker.EXPECT().SendBatch(gomock.Any()).AnyTimes()
+	valPerformance := vmock.NewMockValidatorPerformance(ctrl)
+	valPerformance.EXPECT().ValidatorPerformanceScore(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(num.DecimalFromFloat(1)).AnyTimes()
+
+	validatorData := []*types.ValidatorData{validator1, validator2, validator3, validator4}
+	valScores := map[string]num.Decimal{"node1": num.DecimalFromFloat(0.25), "node2": num.DecimalFromFloat(0.5), "node3": num.DecimalFromFloat(0.25), "node4": num.DecimalZero()}
+	res := calculateRewardsByStake("1", "asset", "rewardsAccountID", num.NewUint(1000000), valScores, validatorData, delegatorShare, num.NewUint(1000000000), logging.NewTestLogger())
+
+	// 0.1% of 1000000000 = 1000000 - this test is demonstrating that regardless of the remaining balance to give to delegators is less than 0.1% of the max
+	// payout per participant, we still run one round and then stop.
+
+	// the normalised scores are as follows (from the test above)
+	// node1 - 0.25
+	// node2 - 0.5
+	// node3 - 0.25
+	// node4 - 0
+	// as node4 has 0 score they get nothing.
+	// given a reward of 1000000,
+	//
+	// node1 and its delegators get 250,000
+	// node2 and its delegators get 500,000
+	// node3 and its delegators get 250,000
+	// with a delegator share of 0.3,
+	// delegators to node1 get 0.3 * 250000 = 75000
+	// party1 gets 0.6 * 75000 = 45000
+	// party2 gets 0.3 * 75000 = 22500
+	// n2 gets 0.1 * 75000 = 7500
+	// node1 gets 175000
+	// node2 gets 1 * 500000 = 500000 + 7500 (from the delegation to node1) = 507500
+	// delegators to node3 get 0.3 * 4/7 * 250000 = 42857
+	// party1 gets 42857
+	// node3 gets 1 - (0.3*4/7) = 207142 -> 207142
+	// node1, node2, party1, party2
+	require.Equal(t, 5, len(res.partyToAmount))
+
+	require.Equal(t, num.NewUint(87857), res.partyToAmount["party1"])
+	require.Equal(t, num.NewUint(22500), res.partyToAmount["party2"])
+	require.Equal(t, num.NewUint(175000), res.partyToAmount["n1"])
+	require.Equal(t, num.NewUint(507500), res.partyToAmount["n2"])
+	require.Equal(t, num.NewUint(207142), res.partyToAmount["n3"])
+	require.Equal(t, num.NewUint(999999), res.totalReward)
+}

